TYP Series and DataFrame currently type-check as hashable

MarcoGorelli opened the pull request:
- [ ] closes #40013 
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

MarcoGorelli committed:
make frame and series non-hashable

mzeitlin11 commented:
@MarcoGorelli does this also take care of #40013?

MarcoGorelli commented:
Thanks @mzeitlin11 - indeed, it looks like they're duplicates, I'll close the new one I'd opened

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -1872,12 +1872,17 @@ def _drop_labels_or_levels(self, keys, axis: int = 0):
+    # https://github.com/python/typeshed/issues/2148#issuecomment-520783318
+    # Incompatible types in assignment (expression has type "None", base class
+    # "object" defined the type as "Callable[[object], int]")
+    __hash__: None = _hash  # type: ignore[assignment]
```
this seems sketchy, but i dont have any better ideas

jreback replied:
i agree this seems like a crazy work-around. we shouldn't be paving the way with mypy incompatiblities.

MarcoGorelli replied:
If it's OK to drop the current custom error message, then would
```
__hash__: None. # type: ignore[assignment]
```
(as suggested https://github.com/python/typeshed/issues/2148#issuecomment-520783318) be OK? If so, I'll do that later

MarcoGorelli commented:
@jbrockmendel @jreback I've updated to just use `__hash__: None`, which is what they do in typeshed (https://github.com/python/typeshed/blob/3fa5988a2ac57cf8ce56e2de171a62b39952f96c/stdlib/builtins.pyi#L602)

The builtin error message will now be shown instead, so I've updated a few tests accordingly:
```python
In [1]: hash(pd.Series([1, 2, 3]))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-1f0c9bc1928b> in <module>
----> 1 hash(pd.Series([1, 2, 3]))

TypeError: unhashable type: 'Series'
```

jbrockmendel commented:
no objection here

jreback commented:
ok this seems reasonable. can you add a release note about the changing error message & rebase. ping on green.

simonjayhawkins commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -612,6 +612,7 @@ Other API changes
 - Partially initialized :class:`CategoricalDtype` (i.e. those with ``categories=None`` objects will no longer compare as equal to fully initialized dtype objects.
 - Accessing ``_constructor_expanddim`` on a :class:`DataFrame` and ``_constructor_sliced`` on a :class:`Series` now raise an ``AttributeError``. Previously a ``NotImplementedError`` was raised (:issue:`38782`)
 - Added new ``engine`` and ``**engine_kwargs`` parameters to :meth:`DataFrame.to_sql` to support other future "SQL engines". Currently we still only use ``SQLAlchemy`` under the hood, but more engines are planned to be supported such as ``turbodbc`` (:issue:`36893`)
+- Calling ``hash`` on non-hashable pandas objects will now raise ``TypeError`` with the built-in error message (e.g. ``unhashable type: 'Series'``). Previously it would raise a custom message such as ``"'Series' objects are mutable, thus they cannot be hashed"`` (:issue:`40013`)
```
nit.

```suggestion
- Calling ``hash`` on non-hashable pandas objects will now raise ``TypeError`` with the built-in error message (e.g. ``unhashable type: 'Series'``). Previously it would raise a custom message such as ``'Series' objects are mutable, thus they cannot be hashed`` (:issue:`40013`)
```

simonjayhawkins commented on pandas/core/frame.py:
```diff
@@ -6152,26 +6152,26 @@ def f(vals) -> tuple[np.ndarray, int]:
             return labels.astype("i8", copy=False), len(shape)
 
         if subset is None:
-            subset = self.columns
+            subset_iterable: Iterable = self.columns
```
another nit, preference for
```
        subset_iterable: Iterable
        if subset is None:
            subset_iterable = self.columns
        elif (
```
to help distinguish cases were we need the variable type annotation (normally where the call to a function returns Any and would be redundant in the future when the return type of the called function is typed) vs a wider type used for a variable than the inferred type( i.e. the return type of the initial assignment to a variable)

simonjayhawkins reviewed (commented):
Thanks @MarcoGorelli lgtm. do we need tests for Index and EA (or do we have tests and not check the message)?

jreback commented:
@MarcoGorelli some comment and can  you rebase

jreback reviewed (approved):
can u rebase and ping on green

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/master' into non-hashable-series-or-frame

MarcoGorelli committed:
move to v1.4.0 whatsnew, declare type before if

MarcoGorelli committed:
add test for index

MarcoGorelli committed:
test extensionarray

MarcoGorelli committed:
revert pandas/tests/indexes/test_base.py

MarcoGorelli committed:
revert pandas/tests/extension/test_common.py

MarcoGorelli commented:
> (or do we have tests and not check the message)?

They have tests, but it's only Series and DataFrame that had this custom error message ("'Series' objects are mutable, thus they cannot be hashed"), for EA and indexes it was already "unhashable type:" (however, they still type-checked as hashable)

simonjayhawkins reviewed (approved):
Thanks @MarcoGorelli lgtm (pending green as will probably get url fails)

see no harm including in 1.3. (except for typing, only the message has changed and we sometimes don't even add release notes for just message changes unless the exception type changes)

MarcoGorelli committed:
move to 1.3.0

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/master' into non-hashable-series-or-frame

mwaskom commented:
This is more than a message change because previously `isinstance(x, collections.abc.Hashable)` would return `True` and now it will return `False`, right? (That said, the existing behavior is wrong, so IMO fixing this sooner rather than later would be helpful ... I currently have an awkward workaround to deal with it in a way that mypy understands).

MarcoGorelli commented:
> This is more than a message change because previously `isinstance(x, collections.abc.Hashable)` would return `True` and now it will return `False`, right? 

Indeed, you are correct (I wasn't aware that this would affect that):

```python
# here
In [4]: isinstance(Series([1,2,3]), collections.abc.Hashable)
Out[4]: False

#master
In [2]: isinstance(Series([1,2,3]), collections.abc.Hashable)
Out[2]: True
```

> IMO fixing this sooner rather than later would be helpful

Agreed -

simonjayhawkins commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -703,6 +703,7 @@ Other API changes
 - Added new ``engine`` and ``**engine_kwargs`` parameters to :meth:`DataFrame.to_sql` to support other future "SQL engines". Currently we still only use ``SQLAlchemy`` under the hood, but more engines are planned to be supported such as `turbodbc <https://turbodbc.readthedocs.io/en/latest/>`_ (:issue:`36893`)
 - Removed redundant ``freq`` from :class:`PeriodIndex` string representation (:issue:`41653`)
 - :meth:`ExtensionDtype.construct_array_type` is now a required method instead of an optional one for :class:`ExtensionDtype` subclasses (:issue:`24860`)
+- Calling ``hash`` on non-hashable pandas objects will now raise ``TypeError`` with the built-in error message (e.g. ``unhashable type: 'Series'``). Previously it would raise a custom message such as ``'Series' objects are mutable, thus they cannot be hashed`` (:issue:`40013`)
```
@MarcoGorelli do we need to add the change of isintance of collections.abc.Hashable?

simonjayhawkins commented:
@MarcoGorelli can you merge master to resolve conflicts

jreback commented:
@MarcoGorelli if you can merge master

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/master' into non-hashable-series-or-frame

simonjayhawkins commented:
mypy will fail.

I think undo all the changes to pandas/core/frame.py

and just do

```
subset = cast(Sequence[Hashable], self.columns)
```

could maybe add link to https://github.com/pandas-dev/pandas/issues/28770 as related.

simonjayhawkins commented:
> and just do
> 
> ```
>         if subset is None:
>             subset = cast(Sequence[Hashable], self.columns)
>         elif (
> ```

or undo all the changes to frame.py and could simply ignore `pandas/core/frame.py:6183: error: Incompatible types in assignment (expression has type "Index", variable has type "Union[Hashable, Sequence[Hashable], None]")  [assignment]`

and we can remove the ignore once  #28770 is fixed

MarcoGorelli committed:
cast self.columns to Sequence[Hashable]

MarcoGorelli committed:
ignore mypy error, pending on 28770

jreback commented on pandas/core/frame.py:
```diff
@@ -6180,7 +6180,10 @@ def f(vals) -> tuple[np.ndarray, int]:
 
         if subset is None:
-            subset = self.columns
+            # Incompatible types in assignment
```
can eliminate these by either casting (as you do on L6195) or better just to assign to a new variable.

MarcoGorelli replied:
@simonjayhawkins do you have a preference here? In https://github.com/pandas-dev/pandas/pull/41283#issuecomment-868986909 you'd suggested to put this instead of the cast and to remove it after #28770

simonjayhawkins replied:
my preference is for the "fix later" ignore rather than changing code for a potential false positive

jreback reviewed (changes requested):
small comment. ping on green.

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/master' into non-hashable-series-or-frame

jreback commented:
thanks @MarcoGorelli

jreback commented:
@meeseeksdev backport 1.3.x
